File: sample/cs/07_game_of_life_loop.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Pytra.CsModule;

public static class Program
{
    // 07: Sample that outputs Game of Life evolution as a GIF.
    
    public static System.Collections.Generic.List<System.Collections.Generic.List<long>> next_state(System.Collections.Generic.List<System.Collections.Generic.List<long>> grid, long w, long h)
    {
        System.Collections.Generic.List<System.Collections.Generic.List<long>> nxt = new System.Collections.Generic.List<System.Collections.Generic.List<long>>();
        long y = 0;
        for (y = 0; y < h; y += 1) {
            System.Collections.Generic.List<long> row = new System.Collections.Generic.List<long>();
            long x = 0;
            for (x = 0; x < w; x += 1) {
                long cnt = 0;
                long dy = -1;
                for (dy = -1; dy < 2; dy += 1) {
                    long dx = -1;
                    for (dx = -1; dx < 2; dx += 1) {
                        if ((dx != 0) || (dy != 0)) {
                            long nx = (x + dx + w) % w;
                            long ny = (y + dy + h) % h;
                            cnt += Pytra.CsModule.py_runtime.py_get(Pytra.CsModule.py_runtime.py_get(grid, ny), nx);
                        }
                    }
                }
                long alive = Pytra.CsModule.py_runtime.py_get(Pytra.CsModule.py_runtime.py_get(grid, y), x);
                if ((alive == 1) && (((cnt == 2) || (cnt == 3)))) {
                    row.Add(1);
                } else {
                    if ((alive == 0) && (cnt == 3)) {
                        row.Add(1);
                    } else {
                        row.Add(0);
                    }
                }
            }
            nxt.Add(row);
        }
        return nxt;
    }
    
    public static List<byte> render(System.Collections.Generic.List<System.Collections.Generic.List<long>> grid, long w, long h, long cell)
    {
        long width = w * cell;
        long height = h * cell;
        List<byte> frame = Pytra.CsModule.py_runtime.py_bytearray(width * height);
        long y = 0;
        for (y = 0; y < h; y += 1) {
            long x = 0;
            for (x = 0; x < w; x += 1) {
                long v = (Pytra.CsModule.py_runtime.py_get(Pytra.CsModule.py_runtime.py_get(grid, y), x) != 0 ? 255 : 0);
                long yy = 0;
                for (yy = 0; yy < cell; yy += 1) {
                    long py_base = (y * cell + yy) * width + x * cell;
                    long xx = 0;
                    for (xx = 0; xx < cell; xx += 1) {
                        Pytra.CsModule.py_runtime.py_set(frame, py_base + xx, v);
                    }
                }
            }
        }
        return Pytra.CsModule.py_runtime.py_bytes(frame);
    }
    
    public static void run_07_game_of_life_loop()
    {
        long w = 144;
        long h = 108;
        long cell = 4;
        long steps = 105;
        string out_path = "sample/out/07_game_of_life_loop.gif";
        
        double start = Pytra.CsModule.time.perf_counter();
        System.Collections.Generic.List<System.Collections.Generic.List<long>> grid = (new System.Func<System.Collections.Generic.List<System.Collections.Generic.List<long>>>(() => { var __out_5 = new System.Collections.Generic.List<System.Collections.Generic.List<long>>(); foreach (var __it_6 in (new System.Func<System.Collections.Generic.List<long>>(() => { var __out_7 = new System.Collections.Generic.List<long>(); long __start_8 = System.Convert.ToInt64(0); long __stop_9 = System.Convert.ToInt64(h); long __step_10 = System.Convert.ToInt64(1); if (__step_10 == 0) { return __out_7; } if (__step_10 > 0) { for (long __i_11 = __start_8; __i_11 < __stop_9; __i_11 += __step_10) { __out_7.Add(__i_11); } } else { for (long __i_11 = __start_8; __i_11 > __stop_9; __i_11 += __step_10) { __out_7.Add(__i_11); } } return __out_7; }))()) { __out_5.Add((new System.Func<System.Collections.Generic.List<long>>(() => { var __base_1 = new System.Collections.Generic.List<long> { 0 }; long __n_2 = System.Convert.ToInt64(w); if (__n_2 < 0) { __n_2 = 0; } var __out_3 = new System.Collections.Generic.List<long>(); for (long __i_4 = 0; __i_4 < __n_2; __i_4 += 1) { __out_3.AddRange(__base_1); } return __out_3; }))()); } return __out_5; }))();
        
        // Lay down sparse noise so the whole field is less likely to stabilize too early.
        // Avoid large integer literals so all transpilers handle the expression consistently.
        long y = 0;
        for (y = 0; y < h; y += 1) {
            long x = 0;
            for (x = 0; x < w; x += 1) {
                long noise = (x * 37 + y * 73 + x * y % 19 + (x + y) % 11) % 97;
                if (noise < 3) {
                    Pytra.CsModule.py_runtime.py_set(Pytra.CsModule.py_runtime.py_get(grid, y), x, 1);
                }
            }
        }
        // Place multiple well-known long-lived patterns.
        System.Collections.Generic.List<System.Collections.Generic.List<long>> glider = new System.Collections.Generic.List<System.Collections.Generic.List<long>> { new System.Collections.Generic.List<long> { 0, 1, 0 }, new System.Collections.Generic.List<long> { 0, 0, 1 }, new System.Collections.Generic.List<long> { 1, 1, 1 } };
        System.Collections.Generic.List<System.Collections.Generic.List<long>> r_pentomino = new System.Collections.Generic.List<System.Collections.Generic.List<long>> { new System.Collections.Generic.List<long> { 0, 1, 1 }, new System.Collections.Generic.List<long> { 1, 1, 0 }, new System.Collections.Generic.List<long> { 0, 1, 0 } };
        System.Collections.Generic.List<System.Collections.Generic.List<long>> lwss = new System.Collections.Generic.List<System.Collections.Generic.List<long>> { new System.Collections.Generic.List<long> { 0, 1, 1, 1, 1 }, new System.Collections.Generic.List<long> { 1, 0, 0, 0, 1 }, new System.Collections.Generic.List<long> { 0, 0, 0, 0, 1 }, new System.Collections.Generic.List<long> { 1, 0, 0, 1, 0 } };
        
        long gy = 8;
        for (gy = 8; gy < h - 8; gy += 18) {
            long gx = 8;
            for (gx = 8; gx < w - 8; gx += 22) {
                long kind = (gx * 7 + gy * 11) % 3;
                if (kind == 0) {
                    long ph = (glider).Count;
                    long py = 0;
                    for (py = 0; py < ph; py += 1) {
                        long pw = (Pytra.CsModule.py_runtime.py_get(glider, py)).Count;
                        long px = 0;
                        for (px = 0; px < pw; px += 1) {
                            if (Pytra.CsModule.py_runtime.py_get(Pytra.CsModule.py_runtime.py_get(glider, py), px) == 1) {
                                Pytra.CsModule.py_runtime.py_set(Pytra.CsModule.py_runtime.py_get(grid, (gy + py) % h), (gx + px) % w, 1);
                            }
                        }
                    }
                } else {
                    if (kind == 1) {
                        long ph = (r_pentomino).Count;
                        long py = 0;
                        for (py = 0; py < ph; py += 1) {
                            long pw = (Pytra.CsModule.py_runtime.py_get(r_pentomino, py)).Count;
                            long px = 0;
                            for (px = 0; px < pw; px += 1) {
                                if (Pytra.CsModule.py_runtime.py_get(Pytra.CsModule.py_runtime.py_get(r_pentomino, py), px) == 1) {
                                    Pytra.CsModule.py_runtime.py_set(Pytra.CsModule.py_runtime.py_get(grid, (gy + py) % h), (gx + px) % w, 1);
                                }
                            }
                        }
                    } else {
                        long ph = (lwss).Count;
                        long py = 0;
                        for (py = 0; py < ph; py += 1) {
                            long pw = (Pytra.CsModule.py_runtime.py_get(lwss, py)).Count;
                            long px = 0;
                            for (px = 0; px < pw; px += 1) {
                                if (Pytra.CsModule.py_runtime.py_get(Pytra.CsModule.py_runtime.py_get(lwss, py), px) == 1) {
                                    Pytra.CsModule.py_runtime.py_set(Pytra.CsModule.py_runtime.py_get(grid, (gy + py) % h), (gx + px) % w, 1);
                                }
                            }
                        }
                    }
                }
            }
        }
        System.Collections.Generic.List<List<byte>> frames = new System.Collections.Generic.List<List<byte>>();
        long _ = 0;
        for (_ = 0; _ < steps; _ += 1) {
            frames.Add(render(grid, w, h, cell));
            grid = next_state(grid, w, h);
        }
        Pytra.CsModule.gif_helper.save_gif(out_path, w * cell, h * cell, frames, Pytra.CsModule.gif_helper.grayscale_palette());
        double elapsed = Pytra.CsModule.time.perf_counter() - start;
        System.Console.WriteLine(string.Join(" ", new object[] { "output:", out_path }));
        System.Console.WriteLine(string.Join(" ", new object[] { "frames:", steps }));
        System.Console.WriteLine(string.Join(" ", new object[] { "elapsed_sec:", elapsed }));
    }
    
    public static void Main(string[] args)
    {
            run_07_game_of_life_loop();
    }
}

File: sample/java/07_game_of_life_loop.java
final class _07_game_of_life_loop {
    private _07_game_of_life_loop() {
    }


    // 07: Sample that outputs Game of Life evolution as a GIF.

    public static java.util.ArrayList<java.util.ArrayList<Long>> next_state(java.util.ArrayList<java.util.ArrayList<Long>> grid, long w, long h) {
        java.util.ArrayList<java.util.ArrayList<Long>> nxt = new java.util.ArrayList<java.util.ArrayList<Long>>();
        for (long y = 0L; y < h; y += 1L) {
            java.util.ArrayList<Long> row = new java.util.ArrayList<Long>();
            for (long x = 0L; x < w; x += 1L) {
                long cnt = 0L;
                for (long dy = (-(1L)); dy < 2L; dy += 1L) {
                    for (long dx = (-(1L)); dx < 2L; dx += 1L) {
                        if ((((dx) != (0L)) || ((dy) != (0L)))) {
                            long nx = (x + dx + w) % w;
                            long ny = (y + dy + h) % h;
                            cnt += ((Long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((ny) < 0L) ? (((long)(grid.size())) + (ny)) : (ny)))))).get((int)((((nx) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((ny) < 0L) ? (((long)(grid.size())) + (ny)) : (ny)))))).size())) + (nx)) : (nx))))));
                        }
                    }
                }
                long alive = ((Long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).get((int)((((x) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).size())) + (x)) : (x))))));
                if ((((alive) == (1L)) && (((cnt) == (2L)) || ((cnt) == (3L))))) {
                    row.add(1L);
                } else {
                    if ((((alive) == (0L)) && ((cnt) == (3L)))) {
                        row.add(1L);
                    } else {
                        row.add(0L);
                    }
                }
            }
            nxt.add(row);
        }
        return nxt;
    }

    public static java.util.ArrayList<Long> render(java.util.ArrayList<java.util.ArrayList<Long>> grid, long w, long h, long cell) {
        long width = w * cell;
        long height = h * cell;
        java.util.ArrayList<Long> frame = PyRuntime.__pytra_bytearray(width * height);
        for (long y = 0L; y < h; y += 1L) {
            for (long x = 0L; x < w; x += 1L) {
                long v = (((((Long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).get((int)((((x) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).size())) + (x)) : (x)))))) != 0L)) ? (255L) : (0L));
                for (long yy = 0L; yy < cell; yy += 1L) {
                    long base = (y * cell + yy) * width + x * cell;
                    for (long xx = 0L; xx < cell; xx += 1L) {
                        frame.set((int)((((base + xx) < 0L) ? (((long)(frame.size())) + (base + xx)) : (base + xx))), v);
                    }
                }
            }
        }
        return PyRuntime.__pytra_bytearray(frame);
    }

    public static void run_07_game_of_life_loop() {
        long w = 144L;
        long h = 108L;
        long cell = 4L;
        long steps = 105L;
        String out_path = "sample/out/07_game_of_life_loop.gif";
        double start = time.perf_counter();
        java.util.ArrayList<java.util.ArrayList<Long>> grid = new java.util.ArrayList<java.util.ArrayList<Long>>();
        for (long __ = 0L; __ < h; __ += 1L) {
            grid.add(PyRuntime.__pytra_list_repeat(0L, w));
        }
        for (long y = 0L; y < h; y += 1L) {
            for (long x = 0L; x < w; x += 1L) {
                long noise = (x * 37L + y * 73L + x * y % 19L + (x + y) % 11L) % 97L;
                if (((noise) < (3L))) {
                    ((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).set((int)((((x) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)((((y) < 0L) ? (((long)(grid.size())) + (y)) : (y)))))).size())) + (x)) : (x))), 1L);
                }
            }
        }
        java.util.ArrayList<java.util.ArrayList<Long>> glider = new java.util.ArrayList<java.util.ArrayList<Long>>(java.util.Arrays.asList(new java.util.ArrayList<Long>(java.util.Arrays.asList(0L, 1L, 0L)), new java.util.ArrayList<Long>(java.util.Arrays.asList(0L, 0L, 1L)), new java.util.ArrayList<Long>(java.util.Arrays.asList(1L, 1L, 1L))));
        java.util.ArrayList<java.util.ArrayList<Long>> r_pentomino = new java.util.ArrayList<java.util.ArrayList<Long>>(java.util.Arrays.asList(new java.util.ArrayList<Long>(java.util.Arrays.asList(0L, 1L, 1L)), new java.util.ArrayList<Long>(java.util.Arrays.asList(1L, 1L, 0L)), new java.util.ArrayList<Long>(java.util.Arrays.asList(0L, 1L, 0L))));
        java.util.ArrayList<java.util.ArrayList<Long>> lwss = new java.util.ArrayList<java.util.ArrayList<Long>>(java.util.Arrays.asList(new java.util.ArrayList<Long>(java.util.Arrays.asList(0L, 1L, 1L, 1L, 1L)), new java.util.ArrayList<Long>(java.util.Arrays.asList(1L, 0L, 0L, 0L, 1L)), new java.util.ArrayList<Long>(java.util.Arrays.asList(0L, 0L, 0L, 0L, 1L)), new java.util.ArrayList<Long>(java.util.Arrays.asList(1L, 0L, 0L, 1L, 0L))));
        for (long gy = 8L; gy < h - 8L; gy += 18L) {
            for (long gx = 8L; gx < w - 8L; gx += 22L) {
                long kind = (gx * 7L + gy * 11L) % 3L;
                long ph = 0L;
                long pw = 0L;
                long px = 0L;
                long py = 0L;
                if (((kind) == (0L))) {
                    ph = ((long)(glider.size()));
                    for (py = 0L; py < ph; py += 1L) {
                        pw = ((long)(((java.util.ArrayList<Object>)(Object)(glider.get((int)((((py) < 0L) ? (((long)(glider.size())) + (py)) : (py)))))).size()));
                        for (px = 0L; px < pw; px += 1L) {
                            if (((((Long)(((java.util.ArrayList<Object>)(Object)(glider.get((int)((((py) < 0L) ? (((long)(glider.size())) + (py)) : (py)))))).get((int)((((px) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(glider.get((int)((((py) < 0L) ? (((long)(glider.size())) + (py)) : (py)))))).size())) + (px)) : (px))))))) == (1L))) {
                                ((java.util.ArrayList<Object>)(Object)(grid.get((int)(((((gy + py) % h) < 0L) ? (((long)(grid.size())) + ((gy + py) % h)) : ((gy + py) % h)))))).set((int)(((((gx + px) % w) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)(((((gy + py) % h) < 0L) ? (((long)(grid.size())) + ((gy + py) % h)) : ((gy + py) % h)))))).size())) + ((gx + px) % w)) : ((gx + px) % w))), 1L);
                            }
                        }
                    }
                } else {
                    if (((kind) == (1L))) {
                        ph = ((long)(r_pentomino.size()));
                        for (py = 0L; py < ph; py += 1L) {
                            pw = ((long)(((java.util.ArrayList<Object>)(Object)(r_pentomino.get((int)((((py) < 0L) ? (((long)(r_pentomino.size())) + (py)) : (py)))))).size()));
                            for (px = 0L; px < pw; px += 1L) {
                                if (((((Long)(((java.util.ArrayList<Object>)(Object)(r_pentomino.get((int)((((py) < 0L) ? (((long)(r_pentomino.size())) + (py)) : (py)))))).get((int)((((px) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(r_pentomino.get((int)((((py) < 0L) ? (((long)(r_pentomino.size())) + (py)) : (py)))))).size())) + (px)) : (px))))))) == (1L))) {
                                    ((java.util.ArrayList<Object>)(Object)(grid.get((int)(((((gy + py) % h) < 0L) ? (((long)(grid.size())) + ((gy + py) % h)) : ((gy + py) % h)))))).set((int)(((((gx + px) % w) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)(((((gy + py) % h) < 0L) ? (((long)(grid.size())) + ((gy + py) % h)) : ((gy + py) % h)))))).size())) + ((gx + px) % w)) : ((gx + px) % w))), 1L);
                                }
                            }
                        }
                    } else {
                        ph = ((long)(lwss.size()));
                        for (py = 0L; py < ph; py += 1L) {
                            pw = ((long)(((java.util.ArrayList<Object>)(Object)(lwss.get((int)((((py) < 0L) ? (((long)(lwss.size())) + (py)) : (py)))))).size()));
                            for (px = 0L; px < pw; px += 1L) {
                                if (((((Long)(((java.util.ArrayList<Object>)(Object)(lwss.get((int)((((py) < 0L) ? (((long)(lwss.size())) + (py)) : (py)))))).get((int)((((px) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(lwss.get((int)((((py) < 0L) ? (((long)(lwss.size())) + (py)) : (py)))))).size())) + (px)) : (px))))))) == (1L))) {
                                    ((java.util.ArrayList<Object>)(Object)(grid.get((int)(((((gy + py) % h) < 0L) ? (((long)(grid.size())) + ((gy + py) % h)) : ((gy + py) % h)))))).set((int)(((((gx + px) % w) < 0L) ? (((long)(((java.util.ArrayList<Object>)(Object)(grid.get((int)(((((gy + py) % h) < 0L) ? (((long)(grid.size())) + ((gy + py) % h)) : ((gy + py) % h)))))).size())) + ((gx + px) % w)) : ((gx + px) % w))), 1L);
                                }
                            }
                        }
                    }
                }
            }
        }
        java.util.ArrayList<java.util.ArrayList<Long>> frames = new java.util.ArrayList<java.util.ArrayList<Long>>();
        for (long __ = 0L; __ < steps; __ += 1L) {
            frames.add(render(grid, w, h, cell));
            grid = next_state(grid, w, h);
        }
        gif.save_gif(out_path, w * cell, h * cell, frames, gif.grayscale_palette(), 4L, 0L);
        double elapsed = time.perf_counter() - start;
        System.out.println(String.valueOf("output:") + " " + String.valueOf(out_path));
        System.out.println(String.valueOf("frames:") + " " + String.valueOf(steps));
        System.out.println(String.valueOf("elapsed_sec:") + " " + String.valueOf(elapsed));
    }
}
